### PR TITLE
Add Tree View Component

### PR DESCRIPTION
**Pull request by DimalChandrasiri**
Mar-17-2020 16-23-44

### CODE REVIEW

File: modules/react-components/src/tree-view/index.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+ export * from "./tree-view";
+
```
Comment (brionmario) [inline modules/react-components/src/tree-view/index.ts]: Add a new line at the end

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { FunctionComponent, ReactElement, useState, useEffect } from "react";
+import isNil from 'lodash/isNil';
```
Comment (brionmario) [inline modules/react-components/src/tree-view/tree-view.tsx]: Shall we import lodash like `import _ from 'lodash';`.  We have done so in other places.

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+import find from 'lodash/find';
+import get from 'lodash/get';
+import cloneDeep from 'lodash/cloneDeep';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
+
+interface TreeViewProps {
+    data: TreeNode[];
+    depth?: number;
+    deleteElement?: ReactElement;
+    getStyleClassCb?: (node, depth?) => {};
```
Comment (brionmario) [inline modules/react-components/src/tree-view/tree-view.tsx]: Shall we add the missing types for the arguments.

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+export const TreeView: FunctionComponent<TreeViewProps> = (props: TreeViewProps): ReactElement => {
+
+    const [ treeData, setTreeData ] = useState<TreeNode[]>();
+    const [ lastCheckToggledNodeIndex, setLastCheckToggledNodeIndex ] = useState<number>();
+
+    useEffect(() => {
+        setTreeData(cloneDeep(props.data));
+    },[props.data]);
```
Comment (brionmario) [inline modules/react-components/src/tree-view/tree-view.tsx]: It'll be better to destructure props at the begining of the component.

```
const { data } = props;
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+ */
+
+import React, { FunctionComponent, ReactElement, useState, useEffect } from "react";
+import isNil from 'lodash/isNil';
+import isEmpty from 'lodash/isEmpty';
+import isEqual from 'lodash/isEqual';
+import find from 'lodash/find';
+import get from 'lodash/get';
+import cloneDeep from 'lodash/cloneDeep';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
```
Comment (brionmario) [inline modules/react-components/src/tree-view/tree-view.tsx]: Use double quotes.

File: modules/react-components/src/tree-view/index.ts (hunk 1 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/index.ts]: Fixed with commit 3a6a06c

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 2 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: Fixed with commit 3a6a06c

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 3 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: Fixed with commit 3a6a06c

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 4 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: Fixed with commit 3a6a06c

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 5 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: Fixed with commit 3a6a06c

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+                >
+                    {deleteElement}
+                </div>
+            );
+        }
+    }
+
+    const printExpandButton = (node: TreeNode): ReactElement => {
+        const className = node.isExpanded
+            ? ''
```
Comment (inthirakumaaran) [inline modules/react-components/src/tree-view/tree-view.tsx]: Any reason have this check in multiple lines?

File: modules/theme/src/definitions/views/admin-portal.less [hunk 7]
```diff
@@ -268,4 +239,116 @@
   height: 1px;
 }
 
+.wrap {
```
Comment (brionmario) [inline modules/theme/src/definitions/views/admin-portal.less]: Shall we scope these classes?

File: modules/theme/src/definitions/views/admin-portal.less [hunk 8]
```diff
@@ -268,4 +239,116 @@
...
+.tree-label {
+  user-select: none;
+  cursor: pointer;
+  margin-bottom: 0;
+}
+.tree-label input:checked + .checkbox {
+  border-color: #20C2E0;
+}
+.tree-label input:checked + .checkbox svg path {
+  fill: #20C2E0;
```
Comment (brionmario) [inline modules/theme/src/definitions/views/admin-portal.less]: +1 to make this the primary color 😁

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 6 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: This was actually a code migrate from an existing plugin to make it easier to style as the way we want. Any how I have fixed this with commit 8afade7

File: modules/theme/src/definitions/views/admin-portal.less (hunk 7 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: Fixed with commit 8afade7

File: modules/theme/src/definitions/views/admin-portal.less (hunk 8 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: Will discuss this and implement.

File: modules/theme/src/definitions/views/admin-portal.less [hunk 9]
```diff
@@ -115,9 +115,7 @@
 }
 
 // VARIABLES
-@super-treeview-text-color: #000;
-@super-treeview-arrow-color: #000;
-@super-treeview-tree-stem-color: #000;
+@super-treeview-tree-stem-color: #f3f3f3;
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: why are we using term `@super` here ?

File: modules/theme/src/definitions/views/admin-portal.less [hunk 10]
```diff
@@ -115,9 +115,7 @@
...
 
 .super-treeview * {
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: Shall we change this to just `treeview` ? may be you can make this a collection item instead

File: modules/theme/src/definitions/views/admin-portal.less [hunk 11]
```diff
@@ -268,4 +239,134 @@
   height: 1px;
 }
 
+.wrap {
+	cursor: pointer;
+	float: left;
+	margin-left: -3px;
+	margin-right: 5px;
+}
+.arrow {
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: `.arrow` class in the root level apply for all the `.arrow` components. Is this intentional ?

File: modules/theme/src/definitions/views/admin-portal.less [hunk 12]
```diff
@@ -268,4 +239,134 @@
...
+	cursor: pointer;
+	display: inline-block;
+	height: 18px;
+	position: relative;
+	top: 4px;
+	width: 18px;
+	span {
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: need a new-line before the `span`

File: modules/theme/src/definitions/views/admin-portal.less [hunk 13]
```diff
@@ -268,4 +239,134 @@
...
+		-webkit-transition: all 0.2s ease;
+		background-color: #bfbfbf;
+		display: inline-block;
+		height: 0.1rem;
+		position: absolute;
+		top: 0.5rem;
+		transition: all 0.2s ease;
+		width: 0.75rem;
+		&:first-of-type {
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: Need a new line before the `&:first-of-type`

File: modules/theme/src/definitions/views/admin-portal.less (hunk 9 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed with commit 6cb1c41

File: modules/theme/src/definitions/views/admin-portal.less (hunk 10 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed with commit 6cb1c41

File: modules/theme/src/definitions/views/admin-portal.less (hunk 12 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed with commit 6cb1c41 #

File: modules/theme/src/definitions/views/admin-portal.less (hunk 13 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed with commit 6cb1c41

File: modules/theme/src/definitions/views/admin-portal.less [hunk 14]
```diff
@@ -115,54 +115,54 @@
 }
 
 // VARIABLES
-@super-treeview-text-color: #000;
-@super-treeview-arrow-color: #000;
-@super-treeview-tree-stem-color: #000;
+@treeview-tree-stem-color: #f3f3f3;
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: Move variables to variable file

Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed with commit 8d0673e7d9a2fca1bb84b137ada06c5569e79369

**Pull request merged**